### PR TITLE
Adding new DataEntry ledger entry type.

### PR DESCRIPTION
**Pull request by jedmccaleb**
Don't merge yet. 
Should all be working now. But we should discuss with @nullstyle and @MonsieurNicolas to make sure this is how we want to make the change.
Start of a possible implementation of: https://github.com/stellar/stellar-protocol/issues/27

### CODE REVIEW

File: src/ledger/DataFrame.cpp [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+#include "util/types.h"
+#include "util/basen.h"
+
+using namespace std;
+using namespace soci;
+
+namespace stellar
+{
+const char* DataFrame::kSQLCreateStatement1 =
+    "CREATE TABLE accountdata"
```
Comment (nullstyle) [inline src/ledger/DataFrame.cpp]: Probably should add an index on (accountid, dataname)

File: src/ledger/DataFrame.cpp [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+
+bool
+DataFrame::exists(Database& db, LedgerKey const& key)
+{
+    std::string actIDStrKey = PubKeyUtils::toStrKey(key.data().accountID);
+    std::string dataName = key.data().dataName;
+    int exists = 0;
+    auto timer = db.getSelectTimer("data-exists");
+    auto prep =
+        db.getPreparedStatement("SELECT EXISTS (SELECT NULL FROM accountdata "
```
Comment (nullstyle) [inline src/ledger/DataFrame.cpp]: What does the execution plan look like on this?  Any reason not to simply use `SELECT 1 from accountdata WHERE account=:id AND dataname=:s`?

I'm always wary of subqueries, especially when aiming against multiple database engines.

File: src/transactions/ManageDataTests.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+using namespace stellar;
+using namespace stellar::txtest;
+
+typedef std::unique_ptr<Application> appPtr;
+
+// add data
+// change data
+// remove data
+// remove data that isn't there
+// add too much data
```
Comment (nullstyle) [inline src/transactions/ManageDataTests.cpp]: Looks like you haven't written the code for this case.  Is that correct?

File: src/xdr/Stellar-transaction.x [hunk 4]
```diff
@@ -205,6 +206,21 @@ Result: InflationResult
     Result : AccountMergeResult
 */
 
+/* ManageData
+Creates and funds a new account with the specified starting balance.
```
Comment (ara4n) [inline src/xdr/Stellar-transaction.x]: (the comment block needs to be updated from the c+p of CreateAccount)

File: src/ledger/DataFrame.h [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    // database utilities
+    static pointer loadData(AccountID const& accountID, std::string dataName,
+                             Database& db);
+
+
+    static void loadAccountsData(AccountID const& accountID,
+                           std::vector<DataFrame::pointer>& retData,
+                           Database& db);
+
+    // load all offers from the database (very slow)
```
Comment (ara4n) [inline src/ledger/DataFrame.h]: (comment's cargoculted from OfferFrame and needs to be updated)

File: src/ledger/DataFrame.cpp (hunk 2 above)
Comment (jedmccaleb) [inline src/ledger/DataFrame.cpp]: I don't remember why nicolas did it this way but they are all done like this.

File: src/transactions/ManageDataTests.cpp (hunk 3 above)
Comment (jedmccaleb) [inline src/transactions/ManageDataTests.cpp]: it is there. It is the one that checks MANAGE_DATA_LOW_RESERVE

File: src/ledger/DataFrame.cpp [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+
+    std::string actIDStrKey = PubKeyUtils::toStrKey(accountID);
+
+    std::string sql = dataColumnSelector;
+    sql += " WHERE accountid = :id AND dataname = :dataname";
+    auto prep = db.getPreparedStatement(sql);
+    auto& st = prep.statement();
+    st.exchange(use(actIDStrKey));
+    st.exchange(use(dataName));
+
```
Comment (graydon) [inline src/ledger/DataFrame.cpp]: Needs an `st.define_and_bind()` here before it'll work.

Comment (graydon) [inline src/ledger/DataFrame.cpp]: oh never mind I see you do it in `loadData()`

**Comment by graydon** [comment]
Ok, this looks pretty good to me. Aside from needing a clang-format run (lots of random tabs and off indentation) and a little curiosity about why 12 character names / 64 byte values, I'm fine with it.

I'm trying to remember whether we decided a pure addition at the protocol level warranted a protocol number rev or not. It's backward compatible but not forward compatible.

**Comment by jedmccaleb** [comment]
@graydon : " and a little curiosity about why 12 character names / 64 byte values" 
I was just picking something that seemed reasonable. Would you choose different lengths for these?

**Comment by jedmccaleb** [comment]
Any ideas for a better name than "DataEntry"

**Comment by graydon** [comment]
I might make the key string a little longer, just to avoid people fighting over salient words. I dunno, we discussed long names before wrt. domain names also, and you seemed opposed. 64 or 128 bytes seems safer to me but it's not a big deal.

Wrt. DataEntry, maybe NamedValueEntry or something? I don't think DataEntry is too bad. It's pretty clear.

**Comment by raymens** [comment]
Some names: KeyValueEntry, DictionaryEntry

File: src/ledger/DataFrame.cpp (hunk 2 above)
Comment (bartekn) [inline src/ledger/DataFrame.cpp]: Looks like it's missing a `)`.

**Comment by jedmccaleb** [comment]
This changes it to be 64byte string keys and 64byte values.

**Comment by bartekn** [comment]
Opened following issues in directly dependent components:
- [ ] https://github.com/stellar/go-stellar-base/issues/18
- [ ] https://github.com/stellar/js-stellar-base/issues/78
- [ ] https://github.com/stellar/java-stellar-sdk/issues/27
- [ ] https://github.com/stellar/ruby-stellar-base/issues/24
- [ ] https://github.com/stellar/docs/issues/102

**Pull request merged**